contributors/design-proposals/network/service_port_ranges.md

```diff
@@ -0,0 +1,128 @@
+
+
+```go
+type ServiceSpec struct {
+  Ports []ServicePort
+}
+
+type ServicePort struct {
+  // portRangeSize is the contiguous ports number that are exposed by this service. Given port = 80 and portRangeSize = 100, it means ports 80-179, 100 ports in total, will be exposed. In this case, port means the starting port of a range.
+  // portRangeSize is default to be 1 for compatibility.
+  // If nodePort is specified, then nodePort is the starting port of range and nodePort+portRangeSize-1 is the ending port of range.
+  // A service can expose multiple port ranges and each port range should not be overlapped.
+  // Service port range does not support target port remapping, it implies that targetPort should not be specified or should be equal to service port.
+  // +optional
+  PortRangeSize int `json: "portRangeSize,omitempty"`
+}
+```
+
+### Service port ranges example
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: service-port-ranges
+  namespace: default
+spec:
+  ports:
+  - name: range1
+    port: 90
+    protocol: TCP
+    portRangeSize: 100
+    protocol: UDP
+  - name: range2
+    port: 200
+    portRange: 300
+    protocol: TCP
+```
+
+### Kube-proxy implementation limitations
+
+NodePort and ExternalIPs Services don't support port ranges since kube-proxy will hold local ports open for these services(nodeport and "external" IP happens to be an IP that is local to this machine). It's challenging to keep some huge port ranges open on host.
+
+Target port remapping is not supported since both iptables and ipvs does not do port translations in the mean time of doing port ranges.
+
+Userspace proxy mode in both linux and windows platform does not work with port ranges. Kube-proxy running in these modes will log error message if people create a service with `portRangeSize` > 1.
+
+#### iptables mode
+
+iptables `multiport` modules support specifying multiple ports in a single line. Suppose service VIP is `1.2.3.4` and port ranges are `1000-2000` and `3000-4000`, then iptables rules might looks like,
+
+```shell
+iptables -A KUBE-SERVICES -d 1.2.3.4/32 -p tcp -m multiport --dports 1000:2000,3000:4000 -j KUBE-SVC-FOO
+```
+
+Given a nodeport type service and port range is `1000-2000`, the iptables rule might looks like,
+
+```shell
+iptables -A NODE-SERVICES -p tcp -m multiport --dports 1000:2000 --dst-type LOCAL -j KUBE-NODEPORTS
+```
+
+#### IPVS mode
+
+IPVS + FWMARK can do port ranges. For example,
+
+```shell
+iptables -A PREROUTNG -t mangle -d 172.16.52.57 -p tcp --dport 1000:2000,3000:4000 -j MARK --set-mark 1
+
+ipvsadm -A -f 1 -s rr
+ipvsadm -a -f 1 -r 172.16.52.60 -m
+ipvsadm -a -f 1 -r 172.16.52.61 -m
+```
+
+Create a fwmark bit range [0-31], for stipulating the IPVS fwmark value range. It’s configurable and should not collison with masqBit and DropBit. By default, it's [0-13] (suppose masqBit is 14 and DropBit is 15), which means we can generate at most 8K fwmark values and cover 8K port ranges.
```

Comment:
Using a large chunk of the fwmark may interfere with other subsystems... And it seems unlikely that you're going to have 8192 distinct service ranges in effect anyway... I'd recommend a much smaller default range.

Reply:
I'm really not fond of this approach still.  It will require us to allocate in a cluster-consistent way (etcd) or else to checkpoint local state so we can survive kube-proxy restarts (I think?).  And it requires carving off a chunk of the very limited fwmark.

I've asked before but not gotten a great answer - why not just forward the whole IP?  We don't have to deal with fwmark.

Or alternately, why not use ipvs to forward the whole IP and iptables to filter ports we don't want to pass thru (blech, did I just say that?)

Reply:
@thockin 

> why not use ipvs to forward the whole IP and iptables to filter ports we don't want to pass thru

It should work and can be an alternative if you really can't live with the fwmark approach.